extensions/in-memory/policy-store-memory/src/main/java/org/eclipse/dataspaceconnector/policy/store/memory/InMemoryPolicyStore.java

```diff
@@ -0,0 +1,99 @@
+/*
+ *  Copyright (c) 2020-2022 Microsoft Corporation
+ *
+ *  This program and the accompanying materials are made available under the
+ *  terms of the Apache License, Version 2.0 which is available at
+ *  https://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  SPDX-License-Identifier: Apache-2.0
+ *
+ *  Contributors:
+ *       Microsoft Corporation - initial API and implementation
+ *
+ */
+package org.eclipse.dataspaceconnector.policy.store.memory;
+
+import org.eclipse.dataspaceconnector.common.concurrency.LockManager;
+import org.eclipse.dataspaceconnector.common.reflection.ReflectionUtil;
+import org.eclipse.dataspaceconnector.policy.model.Policy;
+import org.eclipse.dataspaceconnector.spi.contract.policy.store.PolicyStore;
+import org.eclipse.dataspaceconnector.spi.query.BaseCriterionToPredicateConverter;
+import org.eclipse.dataspaceconnector.spi.query.Criterion;
+import org.eclipse.dataspaceconnector.spi.query.QuerySpec;
+import org.eclipse.dataspaceconnector.spi.query.SortOrder;
+import org.jetbrains.annotations.Nullable;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+import java.util.function.Predicate;
+import java.util.stream.Stream;
+
+import static org.eclipse.dataspaceconnector.common.reflection.ReflectionUtil.propertyComparator;
+
+/**
+ * An in-memory, threadsafe policy store.
+ * This implementation is intended for testing purposes only.
+ */
+public class InMemoryPolicyStore implements PolicyStore {
+
+    private final LockManager lockManager = new LockManager(new ReentrantReadWriteLock());
+    private final Map<String, Policy> policiesById =  new HashMap<>();
+
+    @Override
+    public @Nullable Policy findById(String policyId) {
+        return lockManager.readLock(() -> policiesById.get(policyId));
+    }
+
+    @Override
+    public Stream<Policy> findAll(QuerySpec spec) {
+        return lockManager.readLock(() -> {
+            var stream = policiesById.values().stream();
+            return applyQuery(spec, stream, Policy.class);
+        });
+    }
+
+    @Override
+    public void save(Policy policy) {
+        lockManager.writeLock(() -> policiesById.put(policy.getUid(), policy));
+    }
+
+    @Override
+    public Policy delete(String policyId) {
+        return lockManager.writeLock(() -> policiesById.remove(policyId));
+    }
+
+    private <T> Stream<T> applyQuery(QuerySpec spec, Stream<T> streamOfAll, Class<?> clazz) {
```

Comment:
nit: Looks like this method is not going to be reused by somewhere else. Here this will always return Stream of Policies so it can be a bit simplified by not using generics here.

But it is more of a coding style so you can decide.

Reply:
indeed, I can simplify it here. It's a left over after splitting the code into 2 PRs - this one and the next PR that introduces the refactoring. In the next Pr this method will be extracted to a class. But to avoid confusion I'll simplify it here. Thanks